rfcs/text/0004_application_service_mounting.md

```diff
@@ -0,0 +1,327 @@
+multiple legacy applications on the same page load due to shared stateful
+modules in `ui/public`.
+
+Instead, the ApplicationService should do a full-page refresh when rendering
+legacy applications. Internally, this will be managed by registering legacy apps
+with the ApplicationService separately and handling those top-level routes by
+starting a full-page refresh rather than a mounting cycle.
+
+## Complete Example
+
+Here is a complete example that demonstrates rendering a React application with
+a full-featured router and code-splitting. Note that using React or any other
+3rd party tools featured here is not required to build a Kibana Application.
+
+```tsx
+// my_plugin/public/application.ts
+
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { BrowserRouter, Route } from 'react-router-dom';
+import loadable from '@loadable/component';
+
+// Apps can choose to load components statically in the same bundle or
+// dynamically when routes are rendered.
+import { HomePage } from './pages';
+const LazyDashboard = loadable(() => import('./pages/dashboard'));
+
+const MyApp = ({ basename }) => (
+  // Setup router's basename from the basename provided from MountContext
+  <BrowserRouter basename={basename}>
+
+    {/* mykibana.com/app/my-app/ */}
+    <Route path="/" exact component={HomePage} />
+
+    {/* mykibana.com/app/my-app/dashboard/42 */}
+    <Route
+      path="/dashboard/:id"
+      render={({ match }) => <LazyDashboard dashboardId={match.params.id} />}
+    />
+
+  </BrowserRouter>,
+);
+
+export function renderApp(context, targetDomElement) {
+  ReactDOM.render(
+    // `context.basename` would be `/app/my-app` in this example.
+    // This exact string is not guaranteed to be stable, always reference
+    // `context.basename`.
+    <MyApp basename={context.basename} />,
+    targetDomElem
+  );
+
+  return () => ReactDOM.unmountComponentAtNode(targetDomElem);
+}
+```
+
+```tsx
+// my_plugin/public/plugin.tsx
+
+export class MyPlugin {
+  setup({ application }) {
+    application.register({
+      id: 'my-app',
+      async mount(context, targetDomElem) {
+        const { renderApp } = await import('./applcation');
+        return renderApp(context, targetDomElement);
+      }
+    });
+  }
+}
+```
+
+## Core Entry Point
+
+Once we can support application routing for new and legacy applications, we
+should create a new entry point bundle that only includes Core and any necessary
+uiExports (hacks for example). This should be served by the backend whenever a
+`/app/<app-id>` request is received for an app that the legacy platform does not
+have a bundle for.
+
+# Drawbacks
+
+- Implementing this will be significant work and requires migrating legacy code
+  from `ui/chrome`
+- Making Kibana a single page application may lead to problems if applications
+  do not clean themselves up properly when unmounted
+- Application `mount` functions will have access to *setup* via the closure. We
```

Comment:
We might be able to work around this by using this instead:

```
class Plugin {
  start(core) {
    core.application.register({
      id: 'foo',
      load: () => import(...)
    })
  }
}
```

We could even use eslint to validate that this function is doing nothing other than importing the module and returning the promise. The application service would then need to call the `mount`/`renderApp` function exported by the app module, and since the module is completely outside the scope of the closure it can't reuse the start context.

Reply:
I considered this approach but some downsides to this:
1. Complicates the bundling requirements for 3rd party plugins. 
2. Forces an unnecessary additional request overhead if your app is tiny.

As the API surface of setup shrinks to only registration APIs, I'm less concerned about this closure issue. I'm going to proceed as is, but open to update this RFC before 8.0 if we start running into issues here.